### PR TITLE
Accumulo 2971: improvements to ChangeSecret Tool

### PR DESCRIPTION
**Pull request by milleruntime**
Created a method to verify user running the tool has write permissions to HDFS and modified the tool to update HDFS before making any changes to zookeeper.

### CODE REVIEW

File: server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java [hunk 1]
```diff
@@ -63,12 +66,15 @@ public static void main(String[] args) throws Exception {
     Instance inst = opts.getInstance();
     if (!verifyAccumuloIsDown(inst, opts.oldPass))
       System.exit(-1);
-    String instanceId = rewriteZooKeeperInstance(inst, opts.oldPass, opts.newPass);
-    updateHdfs(fs, inst, instanceId);
+    if (!verifyHdfsWritePermission(fs))
+      System.exit(-1);
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java]: Might be better to print a message first.

Also (and this is perhaps a minor nit), negative exit codes should not be used. They aren't cross-platform. Also, the special exit codes defined in sysexits.h should be avoided, so best to stick with values `[1,63]`. I'd probably just stick with `1` for this (and the other places in this class, which currently use `-1`).

Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java]: A message telling the user they don't have permissions? That is printed in the catch (AccessControlException ace) in verifyHdfsWritePermission().

Will fix the System.exits.

Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java]: Ah, I missed the message in verifyHdfsWritePermission.

Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java]: Fixed in 69042bd

**Comment by mjwall** [comment]
@milleruntime can you make this against the 1.8 branch?  I cherry picked your commits, but had to make a argument final. You should see it when you go to 1.8.

Couple of comments:
- You left a System.exit(1) in the main method.  I'll annotate the diff
- When I removed that System exit, I got the following
`2016-08-04 17:14:43,704 [start.Main] ERROR: Thread 'org.apache.accumulo.server.util.ChangeSecret' died.
java.lang.NoSuchMethodError: org.apache.hadoop.fs.FileSystem.access(Lorg/apache/hadoop/fs/Path;Lorg/apache/hadoop/fs/permission/FsAction;)V
	at org.apache.accumulo.server.util.ChangeSecret.verifyHdfsWritePermission(ChangeSecret.java:172)
	at org.apache.accumulo.server.util.ChangeSecret.main(ChangeSecret.java:68)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.accumulo.start.Main$2.run(Main.java:157)
	at java.lang.Thread.run(Thread.java:745)
`  Probably because I was using hadoop 2.4.1 though.  I was hoping to see the message in the IOException in verifyHdfsWritePermission.
- Can we move the verifyAccumuloIsDown and verifyHdfsWritePermission above the opts parsing.  There is no need to prompt for the instance secret if we know it is going to fail.

I am going to move this to 1.8.1, it can be a bug fix.  I was thinking it changed behavior and was trying to push it into 1.8.0.

File: server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java [hunk 2]
```diff
@@ -61,14 +64,16 @@ public static void main(String[] args) throws Exception {
     opts.parseArgs(ChangeSecret.class.getName(), argsList.toArray(new String[0]));
     VolumeManager fs = VolumeManagerImpl.get();
     Instance inst = opts.getInstance();
-    if (!verifyAccumuloIsDown(inst, opts.oldPass))
-      System.exit(-1);
-    String instanceId = rewriteZooKeeperInstance(inst, opts.oldPass, opts.newPass);
-    updateHdfs(fs, inst, instanceId);
+    verifyAccumuloIsDown(inst, opts.oldPass);
+    System.exit(1);
```
Comment (mjwall) [inline server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java]: This should be removed

**Comment by milleruntime** [comment]
Yeah I was also thinking the checks should be before the password prompts.
Sounds good.

**Comment by ctubbsii** [comment]
Even if it checks beforehand, it could still fail when it tries to actually make the change, right?

**Comment by milleruntime** [comment]
Yeah if you entered the wrong password an auth error will be thrown.

**Comment by ctubbsii** [comment]
@milleruntime I was thinking more that the filesystem permissions could change between the check and the attempt to make the change.

**Comment by milleruntime** [comment]
@mjwall Did you cherry pick these commits? I don't see them in 1.8 branch.
I am currently addressing your comments with a fix to my last commit.

@ctubbsii <https://github.com/ctubbsii> Do you have an idea of how to get
around this filesystem permission change scenario?

On Fri, Aug 5, 2016 at 12:45 PM, Christopher Tubbs <notifications@github.com

**Comment by mjwall** [comment]
@milleruntime I did not cherry pick since it wasn't ready.

**Comment by ctubbsii** [comment]
@milleruntime No. I just wanted to make sure that I properly understood this check. It seems to be a fast-failure check, but not a strict guarantee of future success of the tool. I think that's fine. It's still an improvement.

**Comment by milleruntime** [comment]
@mjwall addressed your first comment with 4d575f8

**Comment by milleruntime** [comment]
@mjwall your second comment points out that the FileSystem access check I added is dependent on a newer version than yours. The method I used is annotated InterfaceAudience.LimitedPrivate, meaning it shouldn't be used by external applications, so I will replace it with something better.

**Comment by milleruntime** [comment]
@mjwall I moved the permissions check before the password prompt. I can't move verifyAccumuloIsDown before the password prompt since it connects to Zookeeper and requires the password.

**Comment by milleruntime** [comment]
Ready to go if no more comments. Squashed changes into e23005faeae9

**Pull request closed without merging**